flink-yarn/src/test/java/org/apache/flink/yarn/YarnResourceManagerDriverTest.java

```diff
@@ -0,0 +1,518 @@
+			.setRemoveContainerRequestConsumer((request, handler) -> removeContainerRequestFuture.complete(null))
+			.setReleaseAssignedContainerConsumer((ignored1, ignored2) -> releaseAssignedContainerFuture.complete(null))
+			.setUnregisterApplicationMasterConsumer((ignore1, ignore2, ignore3) -> stopAndCleanupClusterFuture.complete(null));
+
+		final Map<String, String> env = new HashMap<>();
+
+		private int containerIdx = 0;
+
+		@Override
+		protected void prepareRunTest() {
+			File root = folder.getRoot();
+			File home = new File(root, "home");
+			env.put(ENV_APP_ID, "foo");
+			env.put(ENV_CLIENT_HOME_DIR, home.getAbsolutePath());
+			env.put(ENV_CLIENT_SHIP_FILES, "");
+			env.put(ENV_FLINK_CLASSPATH, "");
+			env.put(ENV_HADOOP_USER_NAME, "foo");
+			env.put(FLINK_DIST_JAR, new YarnLocalResourceDescriptor(
+				"flink.jar",
+				new Path("/tmp/flink.jar"),
+				0,
+				System.currentTimeMillis(),
+				LocalResourceVisibility.APPLICATION,
+				LocalResourceType.FILE).toString());
+			env.put(ApplicationConstants.Environment.PWD.key(), home.getAbsolutePath());
+		}
+
+		@Override
+		protected void preparePreviousAttemptWorkers() {
+			testingYarnAMRMClientAsyncBuilder.setRegisterApplicationMasterFunction(
+				(ignored1, ignored2, ignored3) -> new TestingRegisterApplicationMasterResponse(() -> Collections.singletonList(testingContainer)));
+		}
+
+		@Override
+		protected void prepareReleaseResource() {
+			addContainerRequestFutures.add(new CompletableFuture<>());
+			addContainerRequestFutures.add(new CompletableFuture<>());
+			testingYarnAMRMClientAsyncBuilder.setAddContainerRequestConsumer((ignored1, ignored2) -> {
+				addContainerRequestFutures.get(addContainerRequestFuturesNumCompleted.getAndIncrement()).complete(null);
+				testingYarnResourceManagerClientFactory.getCallbackHandler()
+					.onContainersAllocated(Collections.singletonList(testingContainer));
+			});
+			resourceEventHandlerBuilder.setOnWorkerTerminatedConsumer((ignore1, ignore2) -> getDriver().requestResource(testingTaskExecutorProcessSpec));
+		}
+
+		@Override
+		protected ResourceManagerDriver<YarnWorkerNode> createResourceManagerDriver() {
+			return new YarnResourceManagerDriver(
+				flinkConfig,
+				new YarnResourceManagerDriverConfiguration(env, null, "localhost:9000"),
+				testingYarnResourceManagerClientFactory,
+				testingYarnNodeManagerClientFactory);
+		}
+
+		@Override
+		protected void validateInitialization() {
+			assertNotNull(testingYarnResourceManagerClientFactory.getTestingYarnAMRMClientAsync());
+			assertNotNull(testingYarnNodeManagerClientFactory.getTestingYarnNMClientAsync());
+			assertThat(testingYarnNodeManagerClientFactory.getTestingYarnNMClientAsync().getServiceState(), is(Service.STATE.STARTED));
+			assertThat(testingYarnResourceManagerClientFactory.getTestingYarnAMRMClientAsync().getServiceState(), is(Service.STATE.STARTED));
```

Comment:
I think it would be better to validate whether `init` and `start` are called on these clients. Currently, we are relying on that the state is updated correctly. The problem is that the testing clients override behaviors of real clients, thus a correct state changes on the testing clients cannot guarantee the correct state changes on the real clients.